src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LspFileBasedProgramWorkspace.cs

```diff
@@ -0,0 +1,323 @@
+
+    // TODO: should we use the LanguageServerProjectSystem here instead?
+    private readonly AsyncBatchingWorkQueue<(Uri uri, SourceText documentText, string languageId)> _projectsToLoadAndReload; // TODO: do we need to introduce a shutdown cancellation token for this?
+
+    private readonly ConcurrentDictionary<string, List<LoadedProject>> _loadedProjects = [];
+    private readonly ILspServices _lspServices;
+    private readonly IFileChangeWatcher _fileChangeWatcher;
+    private readonly IEnumerable<Lazy<IDynamicFileInfoProvider, FileExtensionsMetadata>> _dynamicFileInfoProviders;
+    private readonly ProjectTargetFrameworkManager _projectTargetFrameworkManager;
+    private readonly ILoggerFactory _loggerFactory;
+
+    public bool SupportsMutation => true;
+
+    // ILspLogger rationalize
+    ILogger iLogger;
+
+    public LspFileBasedProgramWorkspace(
+        ILspServices lspServices,
+        IFileChangeWatcher fileChangeWatcher,
+        [ImportMany] IEnumerable<Lazy<IDynamicFileInfoProvider, FileExtensionsMetadata>> dynamicFileInfoProviders,
+        ProjectTargetFrameworkManager projectTargetFrameworkManager,
+        ILoggerFactory loggerFactory,
+        IAsynchronousOperationListenerProvider listenerProvider,
+        HostServices hostServices) : base(hostServices, WorkspaceKind.FileBasedPrograms)
+    {
+        _lspServices = lspServices;
+        _fileChangeWatcher = fileChangeWatcher;
+        _dynamicFileInfoProviders = dynamicFileInfoProviders;
+        _projectTargetFrameworkManager = projectTargetFrameworkManager;
+        _loggerFactory = loggerFactory;
+        _projectsToLoadAndReload = new AsyncBatchingWorkQueue<(Uri uri, SourceText documentText, string languageId)>(
+            TimeSpan.FromMilliseconds(100),
+            LoadOrReloadProjectsAsync,
+            equalityComparer: null,
+            listenerProvider.GetListener(FeatureAttribute.Workspace),
+            CancellationToken.None);
+        iLogger = loggerFactory.CreateLogger<LspFileBasedProgramWorkspace>();
+    }
+
+    public async Task<Document?> AddMiscellaneousDocumentAsync(Uri uri, SourceText documentText, string languageId, ILspLogger logger, CancellationToken cancellationToken)
+    {
+        var documentFilePath = ProtocolConversions.GetDocumentFilePathFromUri(uri);
+        var documentId2 = this.CurrentSolution.GetDocumentIdsWithFilePath(documentFilePath);
+        var document2 = this.CurrentSolution.GetDocument(documentId2.FirstOrDefault());
+        if (document2 != null)
+        {
+            return document2;
+        }
+
+        var languageInfoProvider = _lspServices.GetRequiredService<ILanguageInfoProvider>();
+        if (!languageInfoProvider.TryGetLanguageInformation(uri, languageId, out var languageInformation))
+        {
+            // Only log here since throwing here could take down the LSP server.
+            logger.LogError($"Could not find language information for {uri} with absolute path {documentFilePath}");
```

Comment:
does this reflect a bug in roslyn?  if so, use a Contract.Throw.  Does this reflect a bug in a cller?  If so, throw an argument exception.